c++/src/ColumnWriter.cc

```diff
@@ -468,25 +472,1099 @@ namespace orc {
+    ColumnWriter::recordPosition();
+    dataStream->recordPosition(rowIndexPosition.get());
+  }
+
+  class StringColumnWriter : public ColumnWriter {
+  public:
+    StringColumnWriter(const Type& type,
+                       const StreamsFactory& factory,
+                       const WriterOptions& options);
+
+    virtual void add(ColumnVectorBatch& rowBatch,
+                     uint64_t offset,
+                     uint64_t numValues) override;
+
+    virtual void flush(std::vector<proto::Stream>& streams) override;
+
+    virtual uint64_t getEstimatedSize() const override;
+
+    virtual void getColumnEncoding(
+        std::vector<proto::ColumnEncoding>& encodings) const override;
+
+    virtual void recordPosition() const override;
+
+  protected:
+    std::unique_ptr<RleEncoder> lengthEncoder;
+    std::unique_ptr<AppendOnlyBufferedStream> dataStream;
+    RleVersion rleVersion;
+  };
+
+  StringColumnWriter::StringColumnWriter(
+                          const Type& type,
+                          const StreamsFactory& factory,
+                          const WriterOptions& options) :
+                              ColumnWriter(type, factory, options),
+                              rleVersion(RleVersion_1) {
+    std::unique_ptr<BufferedOutputStream> lengthStream =
+        factory.createStream(proto::Stream_Kind_LENGTH);
+    lengthEncoder = createRleEncoder(std::move(lengthStream),
+                                     false,
+                                     rleVersion,
+                                     memPool);
+    dataStream.reset(new AppendOnlyBufferedStream(
+        factory.createStream(proto::Stream_Kind_DATA)));
+
+    if (enableIndex) {
+      recordPosition();
+    }
+  }
+
+  void StringColumnWriter::add(ColumnVectorBatch& rowBatch,
+                               uint64_t offset,
+                               uint64_t numValues) {
+    ColumnWriter::add(rowBatch, offset, numValues);
+    const StringVectorBatch & stringBatch =
+      dynamic_cast<const StringVectorBatch &>(rowBatch);
+
+    char *const * data = stringBatch.data.data() + offset;
+    const int64_t* length = stringBatch.length.data() + offset;
+    const char* notNull = stringBatch.hasNulls ?
+                          stringBatch.notNull.data() + offset : nullptr;
+
+    lengthEncoder->add(length, numValues, notNull);
+
+    for (uint64_t i = 0; i < numValues; ++i) {
+      if (!notNull || notNull[i]) {
+        dataStream->write(data[i], static_cast<size_t>(length[i]));
+      }
+    }
+
+    StringColumnStatisticsImpl* strStats =
+        dynamic_cast<StringColumnStatisticsImpl*>(colIndexStatistics.get());
+
+    bool hasNull = false;
+    for (uint64_t i = 0; i < numValues; ++i) {
+      if (!notNull || notNull[i]) {
+        strStats->update(data[i],
+                         static_cast<size_t>(length[i]));
+        strStats->increase(1);
+      } else if (!hasNull) {
+        hasNull = true;
```

Comment:
Using a separate `hasNull` variable approach is not being used for the `TimestampColumnWriter`. Will be nice to see be consistent.